File: SundeeFundee/Utilities/WeightCalculations.swift
import Foundation

enum SessionResult {
    case first
    case success
    case failure
}

enum WeightCalculations {
    /// Round to nearest 5 lbs.
    static func roundToNearestFive(_ value: Double) -> Double {
        (value / 5.0).rounded() * 5.0
    }

    /// Calculate target weight based on 1RM and percentage.
    static func calculateTargetWeight(oneRepMax: Double, percentage: Double) -> Double {
        roundToNearestFive(oneRepMax * percentage)
    }

    /// Get next recommended working weight based on session result.
    /// - first: 70% of 1RM
    /// - success: +5 lbs
    /// - failure: -5 lbs (floor at 50% of 1RM)
    static func getNextRecommendedWeight(
        currentWeight: Double,
        result: SessionResult,
        oneRepMax: Double
    ) -> Double {
        switch result {
        case .first:
            return roundToNearestFive(oneRepMax * 0.7)
        case .success:
            return roundToNearestFive(currentWeight + 5)
        case .failure:
            let floor = roundToNearestFive(oneRepMax * 0.5)
            return max(roundToNearestFive(currentWeight - 5), floor)
        }
    }

    /// A set is successful if actualReps >= prescribedReps
    /// AND actualWeight >= prescribedWeight (if prescribed).
    static func wasSetSuccessful(
        actualReps: Int,
        prescribedReps: Int,
        actualWeight: Double,
        prescribedWeight: Double?
    ) -> Bool {
        let repsOk = actualReps >= prescribedReps
        let weightOk = prescribedWeight.map { actualWeight >= $0 } ?? true
        return repsOk && weightOk
    }

    /// Session is successful if ALL sets pass.
    static func wasSessionSuccessful(
        sets: [(actualReps: Int, prescribedReps: Int, actualWeight: Double, prescribedWeight: Double?)]
    ) -> Bool {
        sets.allSatisfy { wasSetSuccessful(actualReps: $0.actualReps, prescribedReps: $0.prescribedReps, actualWeight: $0.actualWeight, prescribedWeight: $0.prescribedWeight) }
    }

    /// Check if a weight represents a new personal record.
    static func isPersonalRecord(weight: Double, previousMax: Double) -> Bool {
        weight > previousMax
    }

    /// Calculate volume load (weight × reps × sets).
    static func calculateVolumeLoad(weight: Double, reps: Int, sets: Int) -> Double {
        weight * Double(reps) * Double(sets)
    }

    /// Plateau detected if last 3 weights vary by less than 5 lbs.
    static func detectPlateau(weights: [Double]) -> Bool {
        guard weights.count >= 3 else { return false }
        let lastThree = Array(weights.suffix(3))
        let maxWeight = lastThree.max() ?? 0
        let minWeight = lastThree.min() ?? 0
        return maxWeight - minWeight < 5
    }
}

File: flutter_app/lib/domain/calculations/weight_calculations.dart
enum SessionResult { first, success, failure }

class WeightCalculations {
  const WeightCalculations._();

  static double roundToNearestFive(double value) {
    return (value / 5.0).roundToDouble() * 5.0;
  }

  static double calculateTargetWeight({
    required double oneRepMax,
    required double percentage,
  }) {
    return roundToNearestFive(oneRepMax * percentage);
  }

  static double getNextRecommendedWeight({
    required double currentWeight,
    required SessionResult result,
    required double oneRepMax,
  }) {
    switch (result) {
      case SessionResult.first:
        return roundToNearestFive(oneRepMax * 0.7);
      case SessionResult.success:
        return roundToNearestFive(currentWeight + 5);
      case SessionResult.failure:
        final double floor = roundToNearestFive(oneRepMax * 0.5);
        return _max(roundToNearestFive(currentWeight - 5), floor);
    }
  }

  static bool wasSetSuccessful({
    required int actualReps,
    required int prescribedReps,
    required double actualWeight,
    required double? prescribedWeight,
  }) {
    final bool repsOk = actualReps >= prescribedReps;
    final bool weightOk =
        prescribedWeight == null || actualWeight >= prescribedWeight;
    return repsOk && weightOk;
  }

  static bool wasSessionSuccessful({
    required List<
            ({
              int actualReps,
              int prescribedReps,
              double actualWeight,
              double? prescribedWeight,
            })>
        sets,
  }) {
    return sets.every((
      ({
        int actualReps,
        int prescribedReps,
        double actualWeight,
        double? prescribedWeight,
      }) set,
    ) {
      return wasSetSuccessful(
        actualReps: set.actualReps,
        prescribedReps: set.prescribedReps,
        actualWeight: set.actualWeight,
        prescribedWeight: set.prescribedWeight,
      );
    });
  }

  static bool isPersonalRecord({
    required double weight,
    required double previousMax,
  }) {
    return weight > previousMax;
  }

  static double calculateVolumeLoad({
    required double weight,
    required int reps,
    required int sets,
  }) {
    return weight * reps * sets;
  }

  static bool detectPlateau({required List<double> weights}) {
    if (weights.length < 3) {
      return false;
    }

    final List<double> lastThree = weights.sublist(weights.length - 3);
    final double maxWeight = lastThree.reduce(_max);
    final double minWeight = lastThree.reduce(_min);

    return maxWeight - minWeight < 5;
  }

  static double _max(double a, double b) => a > b ? a : b;

  static double _min(double a, double b) => a < b ? a : b;
}
